master/internal/api_experiment.go

```diff
@@ -957,67 +985,71 @@ func (a *apiServer) GetExperimentCheckpoints(
 		}
 	}
 
-	resp := &apiv1.GetExperimentCheckpointsResponse{}
-	resp.Checkpoints = []*checkpointv1.Checkpoint{}
-	switch err = a.m.db.QueryProto("get_checkpoints_for_experiment", &resp.Checkpoints, req.Id); {
-	case err == db.ErrNotFound:
-		return nil, status.Errorf(
-			codes.NotFound, "no checkpoints found for experiment %d", req.Id)
-	case err != nil:
-		return nil,
-			errors.Wrapf(err, "error fetching checkpoints for experiment %d from database", req.Id)
+	resp := &apiv1.GetExperimentCheckpointsResponse{Checkpoints: []*checkpointv1.Checkpoint{}}
+
+	checkpoints := []*Checkpoint{}
+	q := db.Bun().NewSelect().
+		Model(&checkpoints).
+		Where("experiment_id = ?", req.Id)
+
+	if len(req.States) > 0 {
+		states := []string{}
+		for _, s := range req.States {
+			state_text := checkpointv1.State_name[int32(s.Number())]
+			states = append(states, state_text)
+		}
+		q.Where("state in (?)", bun.In(states))
+
+	}
+
+	orderColumn := "trial_id, report_time"
+	switch req.SortBy {
+	case apiv1.GetExperimentCheckpointsRequest_SORT_BY_BATCH_NUMBER:
+		orderColumn = "(metadata->>'steps_completed')::int"
+	case apiv1.GetExperimentCheckpointsRequest_SORT_BY_UUID:
+		orderColumn = "uuid"
+	case apiv1.GetExperimentCheckpointsRequest_SORT_BY_END_TIME:
+		orderColumn = "report_time"
+	case apiv1.GetExperimentCheckpointsRequest_SORT_BY_STATE:
+		orderColumn = "state_enum"
+	case apiv1.GetExperimentCheckpointsRequest_SORT_BY_SEARCHER_METRIC:
+		orderColumn = "(training->>'searcher_metric')::float8"
+	case apiv1.GetExperimentCheckpointsRequest_SORT_BY_TRIAL_ID:
+	case apiv1.GetExperimentCheckpointsRequest_SORT_BY_UNSPECIFIED:
+	default:
 	}
 
-	a.filter(&resp.Checkpoints, func(i int) bool {
-		v := resp.Checkpoints[i]
+	sortDirection := db.SortDirectionAsc
+	if req.OrderBy == apiv1.OrderBy_ORDER_BY_DESC {
+		sortDirection = db.SortDirectionDescNullsLast
+	}
 
-		found := false
-		for _, state := range req.States {
-			if state == v.State {
-				found = true
-				break
-			}
-		}
+	if req.Limit == 0 {
+		req.Limit = 100
+	}
 
-		if len(req.States) != 0 && !found {
-			return false
-		}
+	db.PaginateBunUnsafe(
+		q,
+		orderColumn,
+		sortDirection,
+		int(req.Offset),
+		int(req.Limit))
 
-		return true
-	})
+	err = q.Scan(context.TODO(), &checkpoints)
 
-	sort.Slice(resp.Checkpoints, func(i, j int) bool {
-		ai, aj := resp.Checkpoints[i], resp.Checkpoints[j]
-		if useSearcherSortBy {
-			if order, done := protoless.CheckpointSearcherMetricNullsLast(ai, aj); done {
-				return order
-			}
-		}
+	switch {
+	case errors.Cause(err) == sql.ErrNoRows || err == db.ErrNotFound:
+		return nil, status.Errorf(
+			codes.NotFound, "no checkpoints found for experiment %d", req.Id)
```

Comment:
added `sql.ErrNoRows` based on [this conversation](https://github.com/determined-ai/determined/pull/4976#discussion_r966027820)